### PR TITLE
Make the admin accept both the old and the new validator

### PR DESCRIPTION
**Pull request by almacbe**
There is some problems when try to use other bundles (https://github.com/liip/LiipFunctionalTestBundle/issues/178) with this one due to SonataAdmin use in this branch the old validator. Although it is fixed in master and just bring the changes to 2.3 branch because I have to do a lot of work to migrate to next version of sonata admin and I think it is easier bring the changes to this branch.

I don't know if it is the right way to do it, but I propose the PR.

### CODE REVIEW

File: Admin/Admin.php [hunk 1]
```diff
@@ -2540,8 +2546,12 @@ public function id($entity)
     /**
      * {@inheritdoc}
      */
-    public function setValidator(ValidatorInterface $validator)
+    public function setValidator($validator)
```
Comment (greg0ire) [inline Admin/Admin.php]: This is a BC-break. By definition, it cannot be merged in 2.3

Comment (almacbe) [inline Admin/Admin.php]: Ouch! It is true... I don't know but maybe there is a way to avoid this BC and solve the problem...

Comment (greg0ire) [inline Admin/Admin.php]: Yeah : you could have an object, implementing `ValidatorInterface` that would hold a reference to the real validator. A bit ugly, but could work.

Comment (almacbe) [inline Admin/Admin.php]: I dont understand you... How could you accept a different class if the interfaces is of another class?

Comment (greg0ire) [inline Admin/Admin.php]: If the argument to setValidator like this : 

```
class ValidatorWrapper implements ValidatorInterface
{
   /* @var mixed ValidatorInterface|LegacyValidatorInterface */
   private $validator;
}
```

Comment (greg0ire) [inline Admin/Admin.php]: But that would mean being able to modify the argument before to wrap it, and I don't think we can do that.

**Comment by almacbe** [comment]
What do you think about the new proposal?

File: Validator/ValidatorWrapper.php [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    private $validator;
+
+    /**
+     * ValidatorWrapper constructor.
+     * @param mixed ValidatorInterface|LegacyValidatorInterface  $validator
+     */
+    public function __construct($validator)
+    {
+        if (!$validator instanceof ValidatorInterface && !$validator instanceof LegacyValidatorInterface) {
+            throw new \InvalidArgumentException('Argument 1 must be an instance of Symfony\Component\Validator\Validator\ValidatorInterface or Symfony\Component\Validator\ValidatorInterface');
```
Comment (greg0ire) [inline Validator/ValidatorWrapper.php]: Please keep it under 120 characters (use concatenation?)

File: Validator/ValidatorWrapper.php [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+    public function getMetadataFactory()
+    {
+        if (method_exists($this->validator, 'getMetadataFor')) {
+            return $this->validator->getMetadataFor($this->getClass());
+        }
+
+        return $this->validator->getMetadataFactory()->getMetadataFor($this->getClass());
+    }
+}
```
Comment (greg0ire) [inline Validator/ValidatorWrapper.php]: Please configure your editor to add a newline at end of file, it will make this pictogram disappear

**Comment by greg0ire** [comment]
> What do you think about the new proposal?

I think it is BC :)

**Comment by almacbe** [comment]
Then? I am sorry but what does patch label means?

**Comment by greg0ire** [comment]
Having tests would be great though, can you write some with prophecy? Something like this : 

```
$wrappedValidator = $this->prophesize('FQCN\Of`ValidatorInterface1');
$wrappedValidator->validate($arg1, $arg2, …)->willReturn('some value');
$validator = new ValidatorWrapper($wrappedValidator->reveal());
$this->assertSame('some value', $validator->validate($arg1, $arg2, …);
```

Use data providers to avoid repetition.

**Comment by greg0ire** [comment]
> Then? I am sorry but what does patch label means?

It means it is a BC change, that brings no new features (it's a bugfix)

**Comment by almacbe** [comment]
I am sorry :-( but it means are you going to merge into 2.3 branch or in master/2.4 branch? Thanks

**Comment by greg0ire** [comment]
It means we can merge that in 2.3 branch (which your P.R is already on).

**Comment by greg0ire** [comment]
depracted => deprecated

File: Resources/config/validator.xml [hunk 4]
```diff
@@ -11,6 +11,10 @@
 
             <tag name="validator.constraint_validator" alias="sonata.admin.validator.inline" />
         </service>
+
+        <service id="sonata_admin.validator.validator_wrapper" class="Sonata\AdminBundle\Validator\ValidatorWrapper">
```
Comment (greg0ire) [inline Resources/config/validator.xml]: Not sure if it will work, but can you try making your service private with `public="false"`? The less public services we have, the less we will have to worry about BC

File: Tests/Validator/ValidatorWrapperTest.php [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+    protected function tearDown()
+    {
+        $this->prophet->checkPredictions();
+    }
+
+    /**
+     * @expectedException InvalidArgumentException
+     */
+    public function testConstructor() {
+        $wrappedValidator = $this->prophet->prophesize('OtherInterface');
```
Comment (greg0ire) [inline Tests/Validator/ValidatorWrapperTest.php]: You can use `$this->prophesize` directly

File: composer.json [hunk 6]
```diff
@@ -48,7 +48,9 @@
         "jms/translation-bundle": "~1.1",
         "symfony/yaml": "~2.3",
         "sonata-project/intl-bundle": "~2.1",
-        "symfony/phpunit-bridge": "~2.7|~3.0"
+        "symfony/phpunit-bridge": "~2.7|~3.0",
+        "phpunit/phpunit": "^4.8",
+        "phpspec/prophecy": "^1.6"
```
Comment (greg0ire) [inline composer.json]: Please don't touch this. Phpunit has prophecy in itself.

File: Tests/Validator/ValidatorWrapperTest.php [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+{
+    /**
+     * @var Prophet $prophet
+     */
+    private $prophet;
+
+    protected function setup()
+    {
+        $this->prophet = new Prophet();
+    }
```
Comment (greg0ire) [inline Tests/Validator/ValidatorWrapperTest.php]: No need for this

File: Tests/Validator/ValidatorWrapperTest.php [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+
+    protected function tearDown()
+    {
+        $this->prophet->checkPredictions();
+    }
```
Comment (greg0ire) [inline Tests/Validator/ValidatorWrapperTest.php]: No need for this either.

File: Tests/Validator/ValidatorWrapperTest.php [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+
+    /**
+     * @expectedException InvalidArgumentException
```
Comment (greg0ire) [inline Tests/Validator/ValidatorWrapperTest.php]: Please don't use this, but `$this->setExpectedException`

File: Tests/Validator/ValidatorWrapperTest.php [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+
+    public function validatorProvider()
+    {
+        return array(
+            array('Symfony\Component\Validator\ValidatorInterface'),
+            array('Symfony\Component\Validator\Validator\RecursiveValidator'),
+        );
+    }
+}
+
```
Comment (greg0ire) [inline Tests/Validator/ValidatorWrapperTest.php]: Please remove line 126

File: composer.json (hunk 6 above)
Comment (almacbe) [inline composer.json]: thanks, I didn't know it

File: Tests/Validator/ValidatorWrapperTest.php [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+/**
+ * Test for ErrorElement.
+ *
+ * @author Alfonso Machado <email@alfonsomachado.com>
+ *
+ * @group validator-wrapper
+ */
+class ValidatorWrapperTest extends \PHPUnit_Framework_TestCase
+{
+    public function testConstructor() {
```
Comment (greg0ire) [inline Tests/Validator/ValidatorWrapperTest.php]: This will trigger a cs warning. Please type `make cs` and amend your commit with the result.

File: composer.json [hunk 12]
```diff
@@ -47,8 +47,7 @@
     "require-dev": {
         "jms/translation-bundle": "~1.1",
         "symfony/yaml": "~2.3",
-        "sonata-project/intl-bundle": "~2.1",
-        "symfony/phpunit-bridge": "~2.7|~3.0"
+        "sonata-project/intl-bundle": "~2.1"
```
Comment (greg0ire) [inline composer.json]: Why would you change anything here?

File: Validator/ValidatorWrapper.php [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+                'Symfony\Component\Validator\Validator\ValidatorInterface '.
+                'or Symfony\Component\Validator\ValidatorInterface';
+
+            throw new \InvalidArgumentException($message);
+        }
+
+        $this->validator = $validator;
+    }
+
+    public function validate($value, $groups = null, $traverse = false, $deep = false)
```
Comment (core23) [inline Validator/ValidatorWrapper.php]: Please add phpDoc

File: Tests/Validator/ValidatorWrapperTest.php (hunk 11 above)
Comment (almacbe) [inline Tests/Validator/ValidatorWrapperTest.php]: sorry, my mistake

File: composer.json (hunk 12 above)
Comment (almacbe) [inline composer.json]: sorry, my mistake

File: Tests/Validator/ValidatorWrapperTest.php [hunk 14]
```diff
@@ -0,0 +1,114 @@
...
+
+        $this->assertSame($metadataFactory, $validator->getMetadataFactory());
+    }
+
+    public function validatorProvider()
+    {
+        return array(
+            array('Symfony\Component\Validator\ValidatorInterface'),
+            array('Symfony\Component\Validator\Validator\RecursiveValidator'),
+        );
```
Comment (greg0ire) [inline Tests/Validator/ValidatorWrapperTest.php]: Change this to return the new interface when it exists, the old interface otherwise.

Comment (almacbe) [inline Tests/Validator/ValidatorWrapperTest.php]: thanks, you are rigth! We will see...

File: Tests/Validator/ValidatorWrapperTest.php [hunk 15]
```diff
@@ -0,0 +1,120 @@
...
+
+        $this->assertSame($metadataFactory, $validator->getMetadataFactory());
+    }
+
+    public function validatorProvider()
+    {
+        if (class_exists('Symfony\Component\Validator\Validator\RecursiveValidator')) {
+            return array(
+                array('Symfony\Component\Validator\ValidatorInterface'),
+                array('Symfony\Component\Validator\Validator\RecursiveValidator'),
```
Comment (greg0ire) [inline Tests/Validator/ValidatorWrapperTest.php]: What is this `RecursiveValidator`? I think you should maybe make the same check on it (have you seen how the build fails?).

Comment (almacbe) [inline Tests/Validator/ValidatorWrapperTest.php]: It Is the new Validator ( https://github.com/symfony/validator/blob/2.8/Validator/RecursiveValidator.php ) I use it instead of the interface because the class implement the legacy interface and it allows to pass the tests. I don't know if it is the right way to do it...

I have seen and I am going to find some time this week to try to fix them.

Comment (greg0ire) [inline Tests/Validator/ValidatorWrapperTest.php]: Ok, thanks!

**Comment by almacbe** [comment]
I dont know if it could be a solution that I propose ... I have researching and it seems that 2.5 and 2.6 there isn't a class that implement the legacy Validator Interface and IMHO there was a "BC". 

Until in 2.7 RecursiveValidator implement it and add the support for legacy validator. I don't know any way to try to implement the same interface in the wrapper for both Validator.

Could you help me with that or give some feedback? Maybe it is imposible to fix it and it is a BC in sonata admin 2.3 and has to add something like dont support 2.8 or something like that, what do you think?

**Comment by greg0ire** [comment]
I don't understand what you said.

> there isn't a class that implement the legacy Validator Interface

contradicts

> Until in 2.7 RecursiveValidator implement it

<=>

> Until in 2.7 RecursiveValidator implement the legacy Validator Interface

<=>

> in symfony <= 2.7 RecursiveValidator implement the legacy Validator Interface

**Comment by core23** [comment]
Sure we should merge this against the 2.3 branch? Thought this branch is dead.

**Comment by greg0ire** [comment]
> Sure we should merge this against the 2.3 branch? Thought this branch is dead.

Not even sure of that, maybe @almacbe should use his own fork. It's just that it's there, like 2.1 and 2.2 are, so if people want to get things merged in it, why not? We will just not support bugs introduced by that.

**Comment by almacbe** [comment]
I am sorry, I didn't explain ver well. Let me try again.

In 2.5 and 2.6 it was introduced the new RecursiveValidator and it didn't implement the legacy ValidatorInterface. It was in 2.7 when RecursiveValidator start to implement it. This is the reason that I have to add some wired condition in the tests.

On the other hand, if someone use 2.5 and 2.6 with RecursiveValidator the wrapper class will be throw a error because it doesn't exist this method and I don't know a way to implement it.

**Comment by greg0ire** [comment]
You should consider using your own fork (that would drop support for old versions). The only other solution I see is make the admin accept also RecursiveValidator (event if does not implement said interfaces).

**Comment by almacbe** [comment]
but you has to change the interface of Admin::setValidator, like I did in the first commit, and it is a BC, isn't?

**Comment by greg0ire** [comment]
Oh, I forgot that… yes it is : https://3v4l.org/tk0ZY So no choice but to create and use your own fork I guess… sorry.

**Comment by almacbe** [comment]
yes...don't worry. Thank's for all I have learnt a lot doing it ;-)

**Comment by greg0ire** [comment]
Have a look at inline aliases if you want to use your fork. Also, if you do this professionally, create the fork in the name of your company. Good luck!
